### ISSUE TITLE
Use a real text layout engine

### ISSUE THREAD

**Issue by azsn** [zelbrium/gllabel#11]
Pango seems like it could work. It might be able to replace the FreeType dependency for font loading too, I'm not sure. HarfBuzz is probably not sufficient on its own because it *just* does shaping -- no line breaks, bidirectional text, justification, etc.

I'm thinking this should look like a class or some functions which take in some text with formatting (e.g. [Pango markup](https://developer.gnome.org/pango/stable/pango-Markup.html)) and returns an array of rectangles specifying the location where each glyph should be placed. (The rectangles could then be converted to triangle verts and sent off to the graphics card.)

**Comment by azsn** [comment]
@discapes If you're still looking for something to do on this project, perhaps it could be interesting to look into this issue. No obligation, of course! Just if you're interested :)

I used Pango once a while back but I've forgotten most of how it works. It does require glib though, which is bloaty. That might be a problem...

**Comment by discapes** [comment]
Based on ['What Harfbuzz doesn't do'](https://harfbuzz.github.io/what-harfbuzz-doesnt-do.html) ,  maybe ICU would be a good option. I don't think making a markup parser would be hard..

**Comment by khaledhosny** [comment]
ICU is still lower level and requires quite a bit of plumbing and still won't handle things like matching fonts and font fallback. Pango is still a better option if one is looking for something that just works.

**Comment by azsn** [comment]
Thanks @khaledhosny , good to know. I haven't used ICU so I can't confirm, but as someone who [once tried to write a widget toolkit](https://github.com/VeltOS/cmk/tree/master/src) I can confirm that text rendering and handling is a nightmare, so I'd prefer to err on the side of something that "just works" haha.

**Comment by discapes** [comment]
So unless anyone has any alternatives, we would accept Pango and just deal with GLib?

**Comment by azsn** [comment]
Yeah, I think that's fine. If we separate components effectively then it should be possible to swap out which layout backend is used in the future. For example, Core Text on Mac or DirectWrite on Windows.

**Comment by azsn** [comment]
@discapes Hey, did you start working on this at all? No problem if not, I just want to make sure I don't duplicate work. If you haven't, I'm going to start looking into pulling out the basic layout code from `gllabel.cpp` into its own section/file (e.g. `text_to_verts()`-kinda thing) probably with a built in "basic layout" (what I already have, proof-of-concept style) and an optional Pango layout (e.g. perhaps with `#define GLLABEL_USE_PANGO`)

**Comment by discapes** [comment]
Nah, I haven't. I'll go check out the progress later if I'm not busy..
